docs/plutus-report/scripting.tex

```diff
@@ -111,7 +115,127 @@ \subsection{Recursive values}
+
+This does complicate interpretation of \gls{plutus-core} programs, since in order to interpret a program, you also need to know what universe of types it is expecting.
+In practice this is not so bad: for most users (e.g. the \gls{cardano} ledger) there will be a single universe that they use, and from there on they can simply act as though we did not have extensible builtins.
+
+\paragraph{Builtin datatypes}
+
+Adding new types to \gls{plutus-core} is not very useful unless they have functions that can operate on them!
+A particularly common case is where the new type is (morally) a datatype, and should support pattern matching.
+We can support this by also providing a ``matching'' function as a builtin function.\footnote{
+  These functions look just like the ``matchers'' for Scott-encoded datatypes, which is no coincidence.
+}
+
+For example, the matching function for \code{bool} is \code{ifThenElse :: forall a . bool -> a -> a -> a}.
+Matching functions are always polymorphic, which is a key reason why we need to support polymorphic builtin functions (see \cref{sec:polymorphic-builtins}).
+
+\paragraph{Polymorphic builtin types}
+
+It would be nice to support polymorphic builtin types.
+For example, that would allow us to provide optimized implementations of container types such as maps.
+
+This seems feasible in principle, but we have not worked the details out yet.
+
+\subsubsection{Builtin functions}
+
+We could implement extensible builtin functions in a similar way to extensible builtin types: use some universe that catalogues the available builtin functions.
+However, this is increasingly tedious, and we don't have the same dependency on the available functions as we do on the types, since we don't have anything corresponding to constants.
+
+So instead we adopt a much simpler approach where the available builtin functions can be provided at runtime simply indexed by a name.
+
+\paragraph{Passing arguments to builtin functions}
+
+Builtin functions must of course be provided with a type signature (since we have a typed language), and a way to evaluate them.
+The natural way to provide such evaluators is as functions --- but these functions must be able to operate on \gls{plutus-core} terms!
+For example, an evaluator for ``plus'' must be able to take two \emph{terms} of type \code{Integer} and turn them into another term.
+But a simple implementation of ``plus'' just talks about \emph{integers}, not terms.
+
+Solving this problem in generality is difficult, so we make a key simplification: builtin functions may only accept arguments of builtin types (with one exception, as we will see shortly).
+Along with the fact that (since \gls{plutus-core} is eager) function arguments are evaluated to values before being passed to the function, this means that the arguments to builtin evaluators will always be \emph{constants}, which are easy to unwrap into their underlying values.
```

Comment:
I guess we could come up with something in a lazy language as well, but good point.

Reply:
Wouldn't we have to make builtins strict in that case? Probably we could come up with something, though...

Reply:
> Wouldn't we have to make builtins strict in that case?

Sure, just like `(+)` is strict in Haskell.